File: src/sync/engine/coordination.rs
//! 同步周期所有权、请求合并与路径活动门禁。

use super::{AppError, AppResult, SyncEngine};
use parking_lot::Mutex;
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use tokio::sync::broadcast;

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
/// 可按位合并的同步周期请求集合。
pub(super) struct CycleRequest(u32);

impl CycleRequest {
    /// 请求重新扫描本地文件树。
    pub(super) const LOCAL_RESCAN: Self = Self(1 << 0);
    /// 请求增量刷新云端变更。
    pub(super) const CLOUD_INCREMENTAL: Self = Self(1 << 1);
    /// 请求全量重建可信云树。
    pub(super) const CLOUD_FULL: Self = Self(1 << 2);
    /// 请求网络恢复后继续持久传输。
    pub(super) const ONLINE_RECOVERY: Self = Self(1 << 3);
    /// 请求执行启动期收敛。
    pub(super) const STARTUP: Self = Self(1 << 4);
    /// 请求重试失败的同步项。
    pub(super) const RETRY: Self = Self(1 << 5);
    /// 仅重新规划一个 RestartRequired 任务，不接受所有失败的同步项进行重试。
    pub(super) const REPLAN: Self = Self(1 << 6);

    /// 判断当前请求集合是否不含任何工作。
    pub(super) fn is_empty(self) -> bool {
        self.0 == 0
    }

    /// 判断当前集合是否包含指定请求位。
    pub(super) fn contains(self, other: Self) -> bool {
        self.0 & other.0 == other.0
    }
}

impl std::ops::BitOr for CycleRequest {
    /// 按位合并后仍使用周期请求类型。
    type Output = Self;

    /// 合并两组周期请求，不丢失任何边沿。
    fn bitor(self, rhs: Self) -> Self::Output {
        Self(self.0 | rhs.0)
    }
}

/// 每个引擎的所有扫描/恢复工作来源都由唯一 owner 协调。请求在等待所有权之前记录，
/// 因而周期执行期间（或其释放窗口）到达的边沿会保持为 sticky request，
/// 并由当前 owner 或下一个等待者消费。
#[derive(Default)]
pub(super) struct CycleCoordinator {
    state: Mutex<CycleCoordinatorState>,
    owner: tokio::sync::Mutex<()>,
}

#[derive(Default)]
/// 记录待处理位、序列进度与有界失败历史。
struct CycleCoordinatorState {
    pending: u32,
    requested: u64,
    completed: u64,
    expired_result_through: u64,
    failures: Vec<(u64, u64, String)>,
}

impl CycleCoordinator {
    /// 合并周期请求并返回其单调序列号。
    pub(super) fn request(&self, request: CycleRequest) -> u64 {
        let mut state = self.state.lock();
        state.requested = state.requested.wrapping_add(1).max(1);
        state.pending |= request.0;
        state.requested
    }

    /// 异步获取唯一周期 owner 所有权。
    pub(super) async fn lock_owner(&self) -> tokio::sync::MutexGuard<'_, ()> {
        self.owner.lock().await
    }

    /// 取走当前待处理请求并保留最新序列号。
    pub(super) fn take_pending_with_sequence(&self) -> (CycleRequest, u64) {
        let mut state = self.state.lock();
        let request = CycleRequest(state.pending);
        state.pending = 0;
        (request, state.requested)
    }

    /// 将明确未执行的请求位恢复到待处理集合。
    pub(super) fn restore(&self, request: CycleRequest) {
        self.state.lock().pending |= request.0;
    }

    /// 标记截止指定序列的周期结束，并记录失败区间。
    pub(super) fn complete(&self, through: u64, error: Option<&AppError>) {
        let mut state = self.state.lock();
        let previous = state.completed;
        state.completed = state.completed.max(through);
        if let Some(error) = error {
            state
                .failures
                .push((previous.saturating_add(1), through, error.to_string()));
            if state.failures.len() > 128 {
                let excess = state.failures.len() - 128;
                let expired_through = state
                    .failures
                    .iter()
                    .take(excess)
                    .map(|(_, end, _)| *end)
                    .max()
                    .unwrap_or(state.expired_result_through);
                state.expired_result_through = state.expired_result_through.max(expired_through);
                state.failures.drain(..excess);
            }
        }
    }

    /// 在序列已完成时返回其成功或失败结果。
    pub(super) fn result_if_completed(&self, sequence: u64) -> Option<AppResult<()>> {
        let state = self.state.lock();
        if state.completed < sequence {
            return None;
        }
        if sequence <= state.expired_result_through {
            return Some(Err(AppError::generic("同步周期结果历史已过期")));
        }
        let failure = state
            .failures
            .iter()
            .find(|(start, end, _)| *start <= sequence && sequence <= *end)
            .map(|(_, _, message)| message.clone());
        Some(match failure {
            Some(message) => Err(AppError::generic(message)),
            None => Ok(()),
        })
    }

    /// 判断是否无待处理请求且当前没有 owner。
    pub(super) fn is_idle(&self) -> bool {
        self.state.lock().pending == 0 && self.owner.try_lock().is_ok()
    }

    /// 判断是否存在已合并但尚未取走的请求。
    pub(super) fn has_pending(&self) -> bool {
        self.state.lock().pending != 0
    }

    /// 判断是否存在尚未完成结算的请求序列。
    pub(super) fn has_uncompleted_request(&self) -> bool {
        let state = self.state.lock();
        state.requested > state.completed
    }
}

#[derive(Default)]
/// 活动门的接纳状态、计数与路径租约。
struct ActivityState {
    accepting: bool,
    count: usize,
    active_paths: HashMap<String, usize>,
    exclusive_paths: HashSet<String>,
}

/// 追踪普通传输与排他路径操作的生命周期。
pub(super) struct ActivityTracker {
    state: Mutex<ActivityState>,
    idle: tokio::sync::Notify,
}

impl Default for ActivityTracker {
    /// 创建默认接受新活动的空追踪器。
    fn default() -> Self {
        Self {
            state: Mutex::new(ActivityState {
                accepting: true,
                count: 0,
                active_paths: HashMap::new(),
                exclusive_paths: HashSet::new(),
            }),
            idle: tokio::sync::Notify::new(),
        }
    }
}

impl ActivityTracker {
    /// 登记普通活动，并拒绝与排他路径重叠的请求。
    pub(super) fn begin(self: &Arc<Self>, relative_path: Option<&str>) -> AppResult<ActivityGuard> {
        let mut state = self.state.lock();
        if !state.accepting {
            return Err(AppError::generic("同步引擎已停止，拒绝新传输活动"));
        }
        let relative_path = relative_path.map(str::to_string);
        if let Some(path) = relative_path.as_deref() {
            if state
                .exclusive_paths
                .iter()
                .any(|leased| sync_paths_overlap(leased, path))
            {
                return Err(AppError::generic("该路径正在执行破坏性操作，请稍后重试"));
            }
            *state.active_paths.entry(path.to_string()).or_default() += 1;
        }
        state.count += 1;
        Ok(ActivityGuard {
            tracker: self.clone(),
            kind: ActivityKind::Shared(relative_path),
        })
    }

    /// 仅在目标子树无普通或排他活动时获取独占租约。
    pub(super) fn begin_exclusive(
        self: &Arc<Self>,
        relative_path: &str,
    ) -> AppResult<ActivityGuard> {
        let mut state = self.state.lock();
        if !state.accepting {
            return Err(AppError::generic("同步引擎已停止，拒绝破坏性操作"));
        }
        if state
            .active_paths
            .keys()
            .any(|active| sync_paths_overlap(active, relative_path))
            || state
                .exclusive_paths
                .iter()
                .any(|leased| sync_paths_overlap(leased, relative_path))
        {
            return Err(AppError::generic("该路径或其子树存在活动任务，请稍后重试"));
        }
        state.exclusive_paths.insert(relative_path.to_string());
        state.count += 1;
        Ok(ActivityGuard {
            tracker: self.clone(),
            kind: ActivityKind::Exclusive(relative_path.to_string()),
        })
    }

    /// 关闭活动门禁以拒绝新活动；已经登记的活动仍可继续结算并释放 guard。
    pub(super) fn close(&self) {
        self.state.lock().accepting = false;
    }

    /// 等待关闭屏障之前已经登记的全部活动释放。
    pub(super) async fn wait_idle(&self) {
        loop {
            let notified = self.idle.notified();
            if self.state.lock().count == 0 {
                return;
            }
            notified.await;
        }
    }
}

/// 在作用域结束时自动释放活动计数与路径租约。
pub(crate) struct ActivityGuard {
    tracker: Arc<ActivityTracker>,
    kind: ActivityKind,
}

/// 区分共享路径活动与独占子树租约。
enum ActivityKind {
    Shared(Option<String>),
    Exclusive(String),
}

/// 判断两个相对路径是否相同或存在祖先关系。
fn sync_paths_overlap(left: &str, right: &str) -> bool {
    left == right
        || left
            .strip_prefix(right)
            .is_some_and(|suffix| suffix.starts_with('/'))
        || right
            .strip_prefix(left)
            .is_some_and(|suffix| suffix.starts_with('/'))
}

/// 在作用域结束时自动释放目录同步门禁。
pub(crate) struct FolderSyncGuard {
    pub(super) engine: Arc<SyncEngine>,
}

impl Drop for FolderSyncGuard {
    /// 释放目录同步标记。
    fn drop(&mut self) {
        self.engine.end_folder_sync();
    }
}

impl Drop for ActivityGuard {
    /// 释放对应租约，并在所有活动结束时唤醒等待者。
    fn drop(&mut self) {
        let mut state = self.tracker.state.lock();
        match &self.kind {
            ActivityKind::Shared(Some(path)) => {
                if let Some(count) = state.active_paths.get_mut(path) {
                    *count = count.saturating_sub(1);
                    if *count == 0 {
                        state.active_paths.remove(path);
                    }
                }
            }
            ActivityKind::Exclusive(path) => {
                state.exclusive_paths.remove(path);
            }
            ActivityKind::Shared(None) => {}
        }
        state.count = state.count.saturating_sub(1);
        if state.count == 0 {
            self.tracker.idle.notify_waiters();
        }
    }
}

/// 将 Engine 活动追踪器适配为 TaskRunner 活动门。
pub(super) struct TaskRunnerActivityGate(pub(super) Arc<ActivityTracker>);

impl crate::sync::task_runner::TaskActivityGate for TaskRunnerActivityGate {
    /// 为传输任务登记活动并返回自动释放的 guard。
    fn begin(&self, relative_path: Option<&str>) -> AppResult<Box<dyn Send>> {
        Ok(Box::new(self.0.begin(relative_path)?))
    }
}

/// 监听网络转换，在在线边沿或滞后收敛时请求恢复。
pub(super) async fn network_listener_loop<L, R>(
    mut transitions: broadcast::Receiver<crate::core::net_guard::NetworkTransition>,
    is_online: L,
    mut shutdown: tokio::sync::watch::Receiver<bool>,
    reconcile_initial: bool,
    mut request_recovery: R,
) where
    L: Fn() -> bool,
    R: FnMut(),
{
    if *shutdown.borrow() {
        return;
    }
    // 调用方先创建订阅，再读取当前网络 level。此处对 level 做一次收敛，确保 listener
    // 启动前已经发送的 Online 边沿不会成为继续推进所必需的条件。
    if reconcile_initial && is_online() {
        request_recovery();
    }
    loop {
        tokio::select! {
            changed = shutdown.changed() => {
                if changed.is_err() || *shutdown.borrow() {
                    break;
                }
            }
            transition = transitions.recv() => {
                match transition {
                    Ok(crate::core::net_guard::NetworkTransition::Online) => {
                        if is_online() {
                            request_recovery();
                        }
                    }
                    Ok(crate::core::net_guard::NetworkTransition::Offline) => {}
                    Err(broadcast::error::RecvError::Lagged(skipped)) => {
                        tracing::warn!(skipped, "网络转换 listener 滞后，按当前 level 收敛");
                        if is_online() {
                            request_recovery();
                        }
                    }
                    Err(broadcast::error::RecvError::Closed) => break,
                }
            }
        }
    }
}

/// 长期存活的 watcher receiver。广播 Lagged 表示路径级细节已经丢失，因此唯一安全的
/// 收敛动作是请求一次合并后的完整重扫；它不同于 Closed，不能终止 listener，
/// 因为后续文件系统批次仍是权威触发信号。
pub(super) async fn watcher_listener_loop<R>(
    mut changes: broadcast::Receiver<crate::mount::local_watcher::ChangeSet>,
    mut shutdown: tokio::sync::watch::Receiver<bool>,
    mut request_rescan: R,
) where
    R: FnMut(),
{
    if *shutdown.borrow() {
        return;
    }
    loop {
        tokio::select! {
            changed = shutdown.changed() => {
                if changed.is_err() || *shutdown.borrow() {
                    break;
                }
            }
            change = changes.recv() => {
                match change {
                    Ok(_) => request_rescan(),
                    Err(broadcast::error::RecvError::Lagged(skipped)) => {
                        tracing::warn!(skipped, "watcher listener 滞后，请求完整补偿扫描");
                        request_rescan();
                    }
                    Err(broadcast::error::RecvError::Closed) => break,
                }
            }
        }
    }
}

File: lib/service/sync/engine/coordination.dart
/// 引擎协调原语 —— 周期请求位 / 周期协调器 / 活动追踪器。
///
/// 严格对齐 Rust 原版 `src/sync/engine/coordination.rs`：
/// - CycleRequest 7 个触发位，sticky 位或合并，单调序列号
/// - CycleCoordinator 唯一 owner drain：正在跑时新触发合并进 pending，
///   不丢弃也不排队等待执行；失败记录 (first, last, msg) 区间（上限 128）
/// - ActivityTracker 活动门：普通/排他路径租约，close 后拒绝新活动，
///   waitIdle 等待全部登记活动释放
library;

import 'dart:async';

import 'package:petal_link/core/error/app_error.dart';

/// 周期请求位（对齐 Rust `CycleRequest`）。
class CycleRequest {
  /// 本地重扫
  static const int localRescan = 1 << 0;

  /// 云端增量
  static const int cloudIncremental = 1 << 1;

  /// 云端全量
  static const int cloudFull = 1 << 2;

  /// 在线恢复
  static const int onlineRecovery = 1 << 3;

  /// 启动恢复
  static const int startup = 1 << 4;

  /// 全局重试
  static const int retry = 1 << 5;

  /// 重规划
  static const int replan = 1 << 6;

  /// 位掩码
  final int bits;

  const CycleRequest(this.bits);

  /// 组合多个位的工厂
  factory CycleRequest.of(Iterable<int> parts) {
    var bits = 0;
    for (final part in parts) {
      bits |= part;
    }
    return CycleRequest(bits);
  }

  /// 是否为空请求
  bool get isEmpty => bits == 0;

  /// 是否包含 [other] 的全部位
  bool contains(int other) => bits & other == other;

  /// 位或合并
  CycleRequest operator |(CycleRequest other) =>
      CycleRequest(bits | other.bits);

  @override
  String toString() => 'CycleRequest(0x${bits.toRadixString(2)})';
}

/// 周期协调器（对齐 Rust `CycleCoordinator`）。
///
/// 防重入语义：正在跑时新触发以位或方式合并进 pending（sticky），
/// 并分配单调递增序列号；owner 的 drain 循环每轮取走全部位。
/// Dart 单 isolate 下 owner 锁用 Future 链互斥实现。
class CycleCoordinator {
  /// 失败历史记录上限
  static const int failuresCap = 128;

  int _pending = 0;
  int _requested = 0;
  int _completed = 0;
  int _expiredResultThrough = 0;

  /// 失败区间（first, last, message）
  final List<(int, int, String)> _failures = [];

  /// owner 互斥锁（Future 链）
  Future<void> _ownerLock = Future<void>.value();

  /// owner 锁是否被持有
  bool _ownerHeld = false;

  /// 合并请求位并分配序列号（序列号从 1 开始，0 表示「无」）。
  int request(CycleRequest req) {
    _requested = (_requested + 1) & 0x7FFFFFFFFFFFFFFF;
    if (_requested < 1) _requested = 1;
    _pending |= req.bits;
    return _requested;
  }

  /// 取 owner 锁（返回释放闭包；同一时刻仅一个 owner 可 drain）。
  Future<void Function()> lockOwner() {
    final completer = Completer<void>();
    final previous = _ownerLock;
    var released = false;
    _ownerLock = previous.then((_) => completer.future);
    return previous.then((_) {
      _ownerHeld = true;
      return () {
        if (released) return;
        released = true;
        _ownerHeld = false;
        completer.complete();
      };
    });
  }

  /// 取走全部 pending 位并返回当前序列号。
  (CycleRequest, int) takePendingWithSequence() {
    final bits = _pending;
    _pending = 0;
    return (CycleRequest(bits), _requested);
  }

  /// 把请求位或回 pending（门控保留语义：shutdown 丢弃、
  /// folderSyncing/离线/云刷新失败/不可信则保留）。
  void restore(CycleRequest req) {
    _pending |= req.bits;
  }

  /// 标记序列号区间已结算；失败时记录区间（超出上限时把挤出区间的
  /// 最大 end 并入 expiredResultThrough）。
  void complete(int through, [Object? error]) {
    if (through > _completed) _completed = through;
    if (error == null) return;
    final first = through; // 单序列区间（drain 每轮一个序列）
    _failures.add((first, through, '$error'));
    while (_failures.length > failuresCap) {
      final evicted = _failures.removeAt(0);
      if (evicted.$2 > _expiredResultThrough) {
        _expiredResultThrough = evicted.$2;
      }
    }
  }

  /// 查询序列号是否已结算（对齐 Rust `result_if_completed`）。
  ///
  /// 返回 null = 未结算；否则返回错误（null 错误 = 成功）。
  /// 历史已过期抛 [StateError]。
  ({bool settled, Object? error}) resultIfCompleted(int sequence) {
    if (_completed < sequence) return (settled: false, error: null);
    if (sequence <= _expiredResultThrough) {
      return (
        settled: true,
        error: AppError.generic('同步周期结果历史已过期'),
      );
    }
    for (final (first, last, message) in _failures) {
      if (sequence >= first && sequence <= last) {
        return (settled: true, error: AppError.generic(message));
      }
    }
    return (settled: true, error: null);
  }

  /// 是否有 pending 请求
  bool hasPending() => _pending != 0;

  /// 是否有未结算请求
  bool hasUncompletedRequest() => _requested > _completed;

  /// 是否空闲（无 pending 且 owner 锁未被持有）
  bool isIdle() => _pending == 0 && !_ownerHeld;
}

/// 路径是否重叠（相等或互为祖先，对齐 Rust `sync_paths_overlap`）。
bool syncPathsOverlap(String left, String right) {
  if (left == right) return true;
  if (left.isEmpty || right.isEmpty) return false;
  return left.startsWith('$right/') || right.startsWith('$left/');
}

/// 活动守卫（RAII 语义；[close] 幂等）。
class ActivityGuard {
  final void Function() _release;
  bool _released = false;

  ActivityGuard(this._release);

  /// 释放活动登记。
  void close() {
    if (_released) return;
    _released = true;
    _release();
  }
}

/// 活动追踪器（对齐 Rust `ActivityTracker`）。
///
/// 普通活动：路径与任一排他租约重叠即拒绝；
/// 排他活动：与任一活动或排他路径重叠即拒绝。
/// [close] 后拒绝新活动（已登记活动仍可结算释放）。
class ActivityTracker {
  bool _accepting = true;
  int _count = 0;
  final Map<String, int> _activePaths = {};
  final Set<String> _exclusivePaths = {};
  final List<Completer<void>> _idleWaiters = [];

  /// 是否仍接受新活动
  bool get isAccepting => _accepting;

  /// 当前登记活动数
  int get count => _count;

  /// 登记一个普通活动（可携带相对路径）。
  ActivityGuard begin([String? path]) {
    if (!_accepting) {
      throw AppError.generic('同步引擎正在停止，拒绝新传输活动');
    }
    if (path != null) {
      for (final exclusive in _exclusivePaths) {
        if (syncPathsOverlap(path, exclusive)) {
          throw AppError.generic('该路径正在执行破坏性操作，请稍后重试');
        }
      }
      _activePaths[path] = (_activePaths[path] ?? 0) + 1;
    }
    _count++;
    return ActivityGuard(() {
      _count--;
      if (path != null) {
        final remaining = (_activePaths[path] ?? 1) - 1;
        if (remaining <= 0) {
          _activePaths.remove(path);
        } else {
          _activePaths[path] = remaining;
        }
      }
      if (_count == 0) _notifyIdle();
    });
  }

  /// 登记一个排他路径活动（破坏性操作用）。
  ActivityGuard beginExclusive(String path) {
    if (!_accepting) {
      throw AppError.generic('同步引擎正在停止，拒绝新传输活动');
    }
    for (final active in _activePaths.keys) {
      if (syncPathsOverlap(path, active)) {
        throw AppError.generic('该路径或其子树存在活动任务，请稍后重试');
      }
    }
    for (final exclusive in _exclusivePaths) {
      if (syncPathsOverlap(path, exclusive)) {
        throw AppError.generic('该路径或其子树存在活动任务，请稍后重试');
      }
    }
    _exclusivePaths.add(path);
    _count++;
    return ActivityGuard(() {
      _count--;
      _exclusivePaths.remove(path);
      if (_count == 0) _notifyIdle();
    });
  }

  /// 拒绝新活动（已登记活动仍可释放）。
  void close() {
    _accepting = false;
    if (_count == 0) _notifyIdle();
  }

  /// 等待全部登记活动释放。
  Future<void> waitIdle() {
    if (_count == 0) return Future<void>.value();
    final completer = Completer<void>();
    _idleWaiters.add(completer);
    return completer.future;
  }

  void _notifyIdle() {
    final waiters = List<Completer<void>>.of(_idleWaiters);
    _idleWaiters.clear();
    for (final waiter in waiters) {
      if (!waiter.isCompleted) waiter.complete();
    }
  }
}
